### PR TITLE
Split head memory and cpu requests/limits

### PR DESCRIPTION
**Pull request by Bobbins228**
# Issue link

Closes: [RHOAIENG-9259](https://issues.redhat.com/browse/RHOAIENG-9259)
# What changes have been made

* Split the head cpu and memory resources to requests/limits similar to #547 
* Added depreciation warnings to the old vars `head_cpus` and `head_memory`
* Updated head/worker_extended_resource_request to include string values due to failing `get_cluster` method
* Updated notebook WF tests to reflect new parameters
* Updated existing e2e tests with new Parameters
* Added documentation for depreciating variables
# Verification steps

### Notebook server ODH/RHOAI/Local
* Clone this repository with `git clone https://github.com/project-codeflare/codeflare-sdk.git`
* Checkout this PR's branch
* Run `poetry build` - install if needed (`pip install poetry`)
* Run `pip install --force-reinstall dist/codeflare_sdk-0.0.0.dev0-py3-none-any.whl` 
* Restart your notebook kernel
### Testing the depreciating args `head_cpus` and `head_memory`
Follow through the [basic Ray demo](https://github.com/project-codeflare/codeflare-sdk/blob/main/demo-notebooks/guided-demos/0_basic_ray.ipynb). Set the `head_cpus` and `head_memory` parameters to a value of your choosing.
You should get a warning that the parameters are being depreciated and to use the new ones.

The head cpu requests and limits should *both* equate the values you entered for the above.

### Testing the new requests/limits args
In the ClusterConfiguration add the parameters 
* head_cpu_requests
* head_cpu_limits
* head_memory_requests 
* head_memory_limits

Set them to values of your choosing and the head pod of the Ray Cluster should reflect these values.
## Checks
- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

File: src/codeflare_sdk/cluster/config.py [hunk 1]
```diff
@@ -36,8 +36,12 @@ class ClusterConfiguration:
     name: str
     namespace: str = None
     head_info: list = field(default_factory=list)
-    head_cpus: typing.Union[int, str] = 2
-    head_memory: typing.Union[int, str] = 8
+    head_cpu_requests: typing.Union[int, str] = 2
```
Comment (astefanutti) [inline src/codeflare_sdk/cluster/config.py:39]: Would it be possible to check how the official Kubernetes Python client types resource requests and limits?

File: src/codeflare_sdk/cluster/model.py [hunk 2]
```diff
@@ -72,8 +72,10 @@ class RayCluster:
...
     status: RayClusterStatus
-    head_cpus: int
-    head_mem: str
+    head_cpu_requests: int
+    head_cpu_limits: int
+    head_mem_requests: str
+    head_mem_limits: str
     head_gpu: int
     workers: int
     worker_mem_min: str
```
Comment (ChristianZaccaria) [inline src/codeflare_sdk/cluster/model.py:81]: worker_mem_min: str
    worker_mem_max: str

Might be worth changing these to `worker_mem_requests/limits`

File: src/codeflare_sdk/cluster/config.py (hunk 1 above)
Comment (Bobbins228) [reply src/codeflare_sdk/cluster/config.py:39]: The official resource [request/limits](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ResourceRequirements.md) is `dict(str, str)`

The SDK accepts both int and str and converts to str before setting the resources later in `generate_yaml.py`

Comment (astefanutti) [reply src/codeflare_sdk/cluster/config.py:39]: What's the plan to support different kind of accelerators, like AMD GPUs or Intel Gaudi?

Comment (Bobbins228) [reply src/codeflare_sdk/cluster/config.py:39]: Kevin has this PR up for supporting different accelerators #531 I think it is pretty much there we should try get it in before next release

**Review by ChristianZaccaria — COMMENTED** [review]
There is one thing to note, perhaps unrelated to this PR, and is that a user can basically input ANY value in the ClusterConfiguration parameters. I.e., I can set to  head_cpu_requests=True, including head_gpus=True, and that is reflected in the yaml as a bool. I believe this is not the expected behaviour. - Note that this was tested on KinD as my OpenShift cluster isn't working at the moment.

**Comment by Bobbins228** [comment]
@ChristianZaccaria 
This is not expected behaviour at all :(
I can have a look at adding some validation to ensure that the head/worker requests/limits are of the correct type.
Good catch!

**Comment by ChristianZaccaria** [comment]
@Bobbins228 I couldn't get further, but I suppose maybe `cluster.up()` will already capture that and throw an error for using the wrong datatypes. However, you're right, there seems to be no validation when creating the yaml file.

**Comment by Bobbins228** [comment]
@ChristianZaccaria This is insane! It seems you can pretty much set any of the variables to whatever type you like.
I will create a Jira for fixing the validation on all ClusterConfiguration parameters.

**Comment by Bobbins228** [comment]
Applied do not merge label until [RHOAIENG-9259](https://issues.redhat.com/browse/RHOAIENG-9259) is a priority again.

File: docs/cluster-configuration.md [hunk 3]
```diff
@@ -28,3 +30,17 @@ Note: 'quay.io/modh/ray:2.35.0-py39-cu121' is the default image used by the Code
 The `labels={"exampleLabel": "example"}` parameter can be used to apply additional labels to the RayCluster resource.
 
 After creating their `cluster`, a user can call `cluster.up()` and `cluster.down()` to respectively create or remove the Ray Cluster.
+
+
+## Depreciating Parameters
```
Comment (KPostOffice) [inline docs/cluster-configuration.md:35]:
```suggestion
## Deprecating Parameters
```

File: docs/cluster-configuration.md [hunk 4]
```diff
@@ -28,3 +30,17 @@ Note: 'quay.io/modh/ray:2.35.0-py39-cu121' is the default image used by the Code
...
+The following parameters of the `ClusterConfiguration` are being depreciated in release `v0.22.0`. <!-- TODO: When removing depreciated parameters update this statement -->
```
Comment (KPostOffice) [inline docs/cluster-configuration.md:36]:
```suggestion
The following parameters of the `ClusterConfiguration` are being deprecated in release `v0.22.0`. <!-- TODO: When removing depreciated parameters update this statement -->
```

File: docs/cluster-configuration.md [hunk 5]
```diff
@@ -28,3 +30,17 @@ Note: 'quay.io/modh/ray:2.35.0-py39-cu121' is the default image used by the Code
...
+| Depreciated Parameter | Replaced By |
```
Comment (KPostOffice) [inline docs/cluster-configuration.md:37]:
```suggestion
| Deprecated Parameter | Replaced By |
```

**Review by KPostOffice — CHANGES_REQUESTED** [review]
This looks good to me, just some docs changes

**Pull request merged**